src/core/server/elasticsearch/client/retry_unauthorized.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { MaybePromise } from '@kbn/utility-types';
+import { AuthHeaders, KibanaRequest, SetAuthHeaders, isRealRequest } from '../../http';
+import { ScopeableRequest } from '../types';
+import { UnauthorizedError } from './errors';
+
+export interface UnauthorizedErrorHandlerOptions {
+  error: UnauthorizedError;
+  request: KibanaRequest;
+}
+
+export type UnauthorizedErrorHandlerResultType = 'notHandled';
+
+export interface UnauthorizedErrorHandlerResultRetryParams {
+  authHeaders: AuthHeaders;
+}
+
+export interface UnauthorizedErrorHandlerRetryResult
+  extends UnauthorizedErrorHandlerResultRetryParams {
+  type: 'retry';
+}
+
+export interface UnauthorizedErrorHandlerNotHandledResult {
+  type: 'notHandled';
+}
+
+export type UnauthorizedErrorHandlerResult =
+  | UnauthorizedErrorHandlerRetryResult
+  | UnauthorizedErrorHandlerNotHandledResult;
+
+interface UnauthorizedErrorHandlerToolkit {
+  /**
+   * The handler cannot handle the error, or was not able to reauthenticate
+   * */
+  notHandled: () => UnauthorizedErrorHandlerResult;
+  /**
+   * The handle was able to reauthenticate.
+   * @param params
+   */
+  retry: (params: UnauthorizedErrorHandlerResultRetryParams) => UnauthorizedErrorHandlerResult;
+}
+
+export const toolkit: UnauthorizedErrorHandlerToolkit = {
+  notHandled: () => ({ type: 'notHandled' }),
+  retry: ({ authHeaders }) => ({
+    type: 'retry',
+    authHeaders,
+  }),
+};
+
+export type UnauthorizedErrorHandler = (
+  options: UnauthorizedErrorHandlerOptions,
+  toolkit: UnauthorizedErrorHandlerToolkit
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+export type InternalUnauthorizedErrorHandler = (
+  error: UnauthorizedError
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+/**
+ * Converts the public version of `UnauthorizedErrorHandler` to the internal one used by the ES client
+ *
+ * @internal
+ */
+export const createInternalErrorHandler = ({
+  handler,
+  request,
+  setAuthHeaders,
+}: {
+  handler: UnauthorizedErrorHandler;
+  request: ScopeableRequest;
+  setAuthHeaders: SetAuthHeaders;
+}): InternalUnauthorizedErrorHandler => {
+  // we don't want to support 401 retry for fake requests
+  if (!isRealRequest(request)) {
+    return () => toolkit.notHandled();
+  }
+  return async (error) => {
+    try {
+      const result = await handler({ request, error }, toolkit);
+      if (isRetryResult(result)) {
+        setAuthHeaders(request, result.authHeaders);
```

Comment:
`setAuthHeaders` just updates the `authRequestHeaders` entry for the given request with the returned headers.

I initially adapted `AuthHeadersStorage` to hold an observable of the AuthHeaders

```diff
-- private authHeadersCache = new WeakMap<Request, AuthHeaders>();
++ private authHeadersCache = new WeakMap<Request, Observable<AuthHeaders>>();
```

to allow reactively responding to auth headers changes. This would have potentially allowed to have multiple scoped clients bound to the same request to reactively update their headers when any client is updated.

e.g (**not working in current implementation**)

```ts
const client1 = clusterClient.asScoped(request);
const client2 = clusterClient.asScoped(request);

// causing the 401 scenario because of expired access token
const responseReturnedAfterRetry =  await client1.search({});

// client2 should ideally already be updated with the new headers and avoid to retrigger the 401
await client2.search({});
```

However, I couldn't find any other scenario where being able to subscribe to auth headers change would provide any value, so I reverted the change to remove the increased complexity. 

WDYT? Do we have a reason/need to change the `AuthHeadersStorage` implementation to hold an observable instead of only the current value of the auth headers?

Reply:
Considering https://github.com/elastic/kibana/issues/108346#issuecomment-976432828. I'd say we can live with it.